fish-rust/src/fish.rs

```diff
@@ -0,0 +1,877 @@
+//
+// The main loop of the fish program.
+/*
+Copyright (C) 2005-2008 Axel Liljencrantz
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License version 2 as
+published by the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
+*/
+
+use autocxx::prelude::*;
+
+use crate::{
+    ast::Ast,
+    builtins::shared::{
+        BUILTIN_ERR_MISSING, BUILTIN_ERR_UNKNOWN, STATUS_CMD_OK, STATUS_CMD_UNKNOWN,
+    },
+    common::{
+        escape_string, exit_without_destructors, get_executable_path, str2wcstring, wcs2string,
+        EscapeStringStyle, PROFILING_ACTIVE, PROGRAM_NAME,
+    },
+    env::{
+        environment::{env_init, EnvStack, Environment},
+        ConfigPaths, EnvMode,
+    },
+    event::{self, Event},
+    ffi::{self, Repin},
+    flog::{self, activate_flog_categories_by_pattern, set_flog_file_fd, FLOG, FLOGF},
+    function, future_feature_flags as features,
+    history::start_private_mode,
+    parse_constants::{ParseErrorList, ParseErrorListFfi, ParseTreeFlags},
+    parse_tree::{ParsedSource, ParsedSourceRefFFI},
+    path::path_get_config,
+    signal::{signal_clear_cancel, signal_unblock_all},
+    threads::{self, asan_maybe_exit},
+    topic_monitor,
+    wchar::prelude::*,
+    wchar_ffi::{WCharFromFFI, WCharToFFI},
+    wutil::waccess,
+};
+use std::env;
+use std::ffi::{CString, OsStr, OsString};
+use std::fs::File;
+use std::io;
+use std::mem::MaybeUninit;
+use std::os::unix::prelude::*;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
+// FIXME: when the crate is actually called fish and not fish-rust, read this from cargo
+// See: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates
+// for reference
+const PACKAGE_NAME: &str = "fish"; // env!("CARGO_PKG_NAME");
+const DOC_DIR: &str = {
+    match option_env!("DOCDIR") {
+        Some(e) => e,
+        None => "share/doc/fish",
```

Comment:
important to note that for any "real" build DOCDIR will be defined and will not need a default.
However we will probably need it for things like `cargo test` and `rust-analyzer`, it least until we move this to `build.rs`
The definition of `DEFAULT_PATH` has the same issue, see its TODO comment.
Can we clarify that these exist purely to facilitate building by using a value like `"(unused)"` everywhere?

Reply:
yeah I think DEFAULT_PATH as fallback in `setup_path` would make sense. Probably has no practical impliciations but it's a simplification